### ISSUE TITLE
03 demo build error

### ISSUE THREAD

**Issue by cswwp** [BIGBALLON/Caffe2_CIFAR#3]
CMake Error at CMakeLists.txt:23 (FIND_PACKAGE):
  By not providing "FindEigen3.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Eigen3", but
  CMake did not find one.

can't find out Eigen3? how fix it?

**Comment by cswwp** [comment]
> CMake Error at CMakeLists.txt:23 (FIND_PACKAGE):
> By not providing "FindEigen3.cmake" in CMAKE_MODULE_PATH this project has
> asked CMake to find a package configuration file provided by "Eigen3", but
> CMake did not find one.
> 
> can't find out Eigen3? how fix it?
It solves by install eigen, but there is another issuse when make /usr/local/include/caffe2/core/common_cudnn.h:7:19: fatal error: cudnn.h: No such file or directory

**Comment by BIGBALLON** [comment]
@cswwp two ways to fix it:

1. remove ``FIND_PACKAGE(Eigen3 REQUIRED)`` from CMakeLists.txt
2. install Eigen3 (by apt-get is ok)

because caffe2 build with third_party Eigen3 not your Eigen3 installed in system, so this error happend.
PS: pls make sure your caffe2 version is less than 0.4.0. since 0.4.1 and 1.0 changed a lot of code, it can't compile correctly

**Comment by BIGBALLON** [comment]
@cswwp plz let me know, if your problem solved or not, thanks a lot.

**Comment by cswwp** [comment]
I install caffe2 c++  as readme 03 part @ BIGBALLON
# Clone Caffe2's source code from Github repository
git clone https://github.com/pytorch/pytorch.git && cd pytorch
# Checkout the tag v0.4.0
git checkout v0.4.0
# Clone third party dependencies
git submodule update --init

# Create a directory to put Caffe2's build files in
mkdir build && cd build

# Configure Caffe2's build
# This looks for packages on your machine and figures out which functionality
# to include in the Caffe2 installation. The output of this command is very
# useful in debugging.
cmake ..

# Compile, link, and install Caffe2 (-j8 is optional)
sudo make install -j8

**Comment by BIGBALLON** [comment]
@cswwp 

``context.h`` and ``common_cudnn.h`` are in my ``/usr/local/include/caffe2/core/``

```
bg@bg-CGI:/usr/local/include/caffe2/core$ ls
allocator.h             context.h         logging_is_google_glog.h      net_dag.h            plan_executor.h             tensor_int8.h
asan.h                  cudnn_wrappers.h  logging_is_not_google_glog.h  net_dag_utils.h      predictor.h                 THCCachingAllocator.h
blob.h                  db.h              macros.h                      net.h                predictor_utils.h           timer.h
blob_serialization.h    dispatch          memonger.h                    net_simple_async.h   qtensor.h                   transform.h
blob_serializer_base.h  event_cpu.h       module.h                      net_simple.h         qtensor_serialization.h     typeid.h
blob_stats.h            event.h           net_async_base.h              nomnigraph           registry.h                  types.h
common_cudnn.h          flags.h           net_async_dag_gpu.h           numa.h               scope_guard.h               workspace.h
common_gpu.h            graph.h           net_async_gpu_thread_pool.h   observer.h           static_tracepoint_elfx86.h
common.h                hip               net_async_polling.h           operator_gradient.h  static_tracepoint.h
common_omp.h            init.h            net_async_scheduling.h        operator.h           stats.h
context_gpu.h           logging.h         net_async_tracing.h           operator_schema.h    tensor.h
```

maybe check your install path ?? or try to add your installed path to environment?

**Comment by cswwp** [comment]
It seems i need to install cudnn

**Comment by BIGBALLON** [comment]
@cswwp yeap, you need to install both cuda & cudnn

**Comment by cswwp** [comment]
> @cswwp yeap, you need to install both cuda & cudnn
why my cmake shows  Could NOT find Caffe2 (missing:  Caffe2_LIBRARY_GPU) 
i install caffe2 as your readme, and i find caffe2 in /usr/local, how fix it?

**Comment by cswwp** [comment]
@BIGBALLON  I set environment in bashrc, but it still can't find caffe2...............
export Caffe2_LIBRARY_GPU=/usr/local/include/caffe2

**Comment by BIGBALLON** [comment]
@cswwp can you use Caffe2's python Version, try to run demo 1 or demo 2?

**Comment by BIGBALLON** [comment]
@cswwp  maybe you can check the installation guide?
https://gist.github.com/BIGBALLON/c428cb75767d40e92be667ee20c4f0c1

did you add the following lines in your bashrc?

```
echo 'export PYTHONPATH=/usr/local:$PYTHONPATH' >> ~/.bashrc
echo 'export PYTHONPATH=$PYTHONPATH:/home/cgilab/pytorch/build' >> ~/.bashrc
echo 'export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH' >> ~/.bashrc
```

**Comment by cswwp** [comment]
> @cswwp maybe you can check the installation guide?
> https://gist.github.com/BIGBALLON/c428cb75767d40e92be667ee20c4f0c1
> 
> did you add the following lines in your bashrc?
> 
> ```
> echo 'export PYTHONPATH=/usr/local:$PYTHONPATH' >> ~/.bashrc
> echo 'export PYTHONPATH=$PYTHONPATH:/home/cgilab/pytorch/build' >> ~/.bashrc
> echo 'export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH' >> ~/.bashrc
> ```

It solves by re-install caffe2 gpu build, thx for you help

**Comment by cswwp** [comment]
> @cswwp maybe you can check the installation guide?
> https://gist.github.com/BIGBALLON/c428cb75767d40e92be667ee20c4f0c1
> 
> did you add the following lines in your bashrc?
> 
> ```
> echo 'export PYTHONPATH=/usr/local:$PYTHONPATH' >> ~/.bashrc
> echo 'export PYTHONPATH=$PYTHONPATH:/home/cgilab/pytorch/build' >> ~/.bashrc
> echo 'export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH' >> ~/.bashrc
> ```
the demo is great, but when i change shufflenet caffe2 model which correct in python, it shows:alled after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at operator.cc:42] blob != nullptr. op Conv: Encountered a non-existing input blob: 0 

i'm struggling with it, how fix it？

**Comment by BIGBALLON** [comment]
@cswwp can you show your code for me? 

``Encountered a non-existing input blob: 0``

it seems your input blob name is not correct.

**Comment by BIGBALLON** [comment]
@cswwp I mean, plz check your shufflenet genrate script(XXX.py), then change the name of both input and output blob in cpp, which may fix your issue.

**Comment by cswwp** [comment]
@BIGBALLON  I have no idea, and pb net struct as follows, help me 
name: "torch-jit-export_predict"
op {
  input: "0"
  input: "1"
  output: "357"
  name: ""
  type: "Conv"
  arg {
    name: "strides"
    ints: 2
    ints: 2
  }
  arg {
    name: "pads"
    ints: 1
    ints: 1
    ints: 1
    ints: 1
  }
  arg {
    name: "dilations"
    ints: 1
    ints: 1
  }
  arg {
    name: "kernels"
    ints: 3
    ints: 3
  }
  arg {
    name: "group"
    i: 1
  }
}
op {
  input: "357"
  input: "2"
  input: "3"
  input: "4"
  input: "5"
  output: "358"
  name: ""
  type: "SpatialBN"
  arg {
    name: "momentum"
    f: 1.0
  }
  arg {
    name: "epsilon"
    f: 9.999999747378752e-06
  }
  arg {
    name: "is_test"
    i: 1
  }
}
op {
  input: "358"
  output: "359"
  name: ""
  type: "Relu"
}
op {
  input: "359"
  output: "360"
  name: ""
  type: "MaxPool"
  arg {
    name: "strides"
    ints: 2
    ints: 2
  }
  arg {
    name: "kernels"
    ints: 3
    ints: 3
  }
  arg {
    name: "pads"
    ints: 0
    ints: 0
    ints: 0
    ints: 0
  }
}
op {
  input: "360"
  input: "7"
  output: "361"
  name: ""
  type: "Conv"
  arg {
    name: "strides"
    ints: 1
    ints: 1
  }
  arg {
    name: "pads"
    ints: 0
    ints: 0
    ints: 0
    ints: 0
  }
  arg {
    name: "dilations"
    ints: 1
    ints: 1
  }
  arg {
    name: "kernels"
    ints: 1
    ints: 1
  }
  arg {
    name: "group"
    i: 1
  }
}
op {
  input: "361"
  input: "8"
  input: "9"
  input: "10"
  input: "11"
  output: "362"
  name: ""
  type: "SpatialBN"
  arg {
    name: "momentum"
    f: 1.0
  }
......

op {
  input: "727"
  output: "728"
  name: ""
  type: "Relu"
}
op {
  input: "728"
  output: "729"
  name: ""
  type: "AveragePool"
  arg {
    name: "global_pooling"
    i: 1
  }
}
op {
  input: "729"
  output: "731"
  name: ""
  type: "Shape"
}
op {
  input: "731"
  input: "730"
  output: "732"
  name: ""
  type: "Gather"
}
op {
  input: "732"
  output: "734"
  name: ""
  type: "ExpandDims"
  arg {
    name: "dims"
    ints: 0
  }
}
op {
  input: "734"
  input: "735"
  output: "736"
  output: "OC2_DUMMY_724"
  name: ""
  type: "Concat"
  arg {
    name: "axis"
    i: 0
  }
}
op {
  input: "729"
  input: "736"
  output: "737"
  output: "OC2_DUMMY_725"
  name: ""
  type: "Reshape"
}
op {
  input: "737"
  input: "355"
  input: "356"
  output: "738"
  name: ""
  type: "FC"
}
op {
  input: "738"
  output: "OC2_DUMMY_726"
  name: ""
  type: "Softmax"
  arg {
    name: "axis"
    i: 1
  }
}
op {
  input: "OC2_DUMMY_726"
  output: "739"
  name: ""
  type: "Log"
}
device_option {
  device_type: 0
  device_id: 0
}
external_input: "0"
external_input: "1"
external_input: "2"
external_input: "3"
external_input: "4"
external_input: "5"
external_input: "7"
external_input: "8"
external_input: "9"
external_input: "10"
external_input: "11"
external_input: "13"
external_input: "14"
....
external_input: "573"
external_input: "595"
external_input: "570"
external_input: "482"
external_input: "504"
external_input: "529"
external_input: "548"
external_input: "551"
external_output: "739"

**Comment by BIGBALLON** [comment]
@cswwp yeap, you need to check your input blob name.

for example;

for the following pb,

```
op {
  input: "data"
  input: "conv1_5x5_128_w"
  input: "conv1_5x5_128_b"
  output: "conv2"
  type: "Conv"
  arg {
    name: "stride"
    i: 1
  }
  arg {
    name: "pad"
    i: 2
  }
...
...
}
op {
  input: "ip_zw"
  output: "tanh"
  type: "Tanh"
}
external_input: "data"
external_input: "conv1_5x5_128_w"
...
...
external_input: "ip_w"
external_input: "ip_b"
external_output: "tanh"

```

the input data blob name is "data",
so you need to use 
```cpp
#ifdef USE_GPU
    auto data = workSpace.GetBlob("data")->GetMutable<TensorCUDA>();
#else
    auto data = workSpace.GetBlob("data")->GetMutable<TensorCPU>();
#endif
```

the output blob name is "tanh"

so you need to use ``tanh`` instead of ``softmax`` in my code.
```cpp
#ifdef USE_GPU
    auto softmax = TensorCPU(workSpace.GetBlob("tanh")->Get<TensorCUDA>());
#else
    auto softmax = workSpace.GetBlob("tanh")->Get<TensorCPU>();
#endif
```

as for you, input blob seems ``0`` and output blob seems ``739``.
so modify the code then it will work,

BTW, maybe check your model generation code(give your layer name if you can), 
re-generate it & modify cpp code, it should work fine,

I guess that is the cause of the error `` Encountered a non-existing input blob``

**Comment by BIGBALLON** [comment]
@cswwp 

try to modify your code:
```
auto data = workSpace.GetBlob("0")->GetMutable<TensorCUDA>();
```

BTW, plz make sure there are no other mistakes in your code.

**Comment by BIGBALLON** [comment]
@cswwp did you solve the problem? If not, maybe I can help you modify the predict code if you want.
